### ISSUE TITLE
Solution: how to run Cecilia5 on modern macs

### ISSUE THREAD

**Issue by funxiun** [belangeo/cecilia5#119]
Hi

I was fiddling around today to get Cecilia5 working again on my Mac Mini M1.
This is what I found:

**Don't** use the release packages from their website (ajaxsoundstudio.com) but use the source code provided here on Github.

First, open a terminal window and install [homebrew](https://brew.sh) :

`/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"

When homebrew is installed, install these packages:
`brew install git flac libogg libvorbis libopusenc mpg123 lame`

Create a directory called git and enter it:
```
mkdir git
cd git
```

Clone the Cecilia5 source code:
`git clone https://github.com/belangeo/cecilia5.git`

Enter the cecilia5 directory and run Cecilia5:
```
cd cecilia5
./Cecilia5.py
```

In my case, Cecilia5 runs :)
Maybe this can be extended by Mac experts how to add a nice icon etc and to avoid entering the terminal window to use Cecilia5.

Hope this helps!

**Comment by oivindi** [comment]
Running on modern Macs hasn't really been the problem - it seems the OS, in this case Sonoma, has been the problem. Are you running Sonoma? :)

**Comment by julienrobert** [comment]
Hi, I tried your instructions and did not succeed.
I was not able to run python without python3, maybe that's the problem. I had to run
```
pip3 install wxPython
pip3 install pyo
```

When I run python3 cecilia5.py, I get this error:

> Traceback (most recent call last):
  File "/Users/jrls/Downloads/cecilia5-5.4.0/Cecilia5.py", line 25, in <module>
    from Resources.Variables import loadBitmaps
  File "/Users/jrls/Downloads/cecilia5-5.4.0/Resources/Variables.py", line 24, in <module>
    from pyo import pa_get_default_devices_from_host
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pyo/__init__.py", line 27, in <module>
    from .lib import analysis as analysis
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pyo/lib/analysis.py", line 30, in <module>
    from ._core import *
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pyo/lib/_core.py", line 47, in <module>
    from .._pyo import *
ImportError: dlopen(/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pyo/_pyo.cpython-311-darwin.so, 0x0002): Library not loaded: /opt/homebrew/opt/flac/lib/libFLAC.12.dylib
  Referenced from: <E6408A3F-C3F8-30CD-906D-54CA42ADE3E5> /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pyo/libsndfile.1.dylib
  Reason: tried: '/opt/homebrew/opt/flac/lib/libFLAC.12.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/opt/flac/lib/libFLAC.12.dylib' (no such file), '/opt/homebrew/opt/flac/lib/libFLAC.12.dylib' (no such file), '/usr/lib/libFLAC.12.dylib' (no such file, not in dyld cache)

**Comment by julienrobert** [comment]
Ok thanks. 

Now I'm getting these errors 

> Preferences file not found. Using defaults...

/Users/jrls/Downloads/cecilia5-5.4.0/Resources/CeciliaPlot.py:971: SyntaxWarning: "is not" with a literal. Did you mean "!="?
  if self._xSpec is not 'none':
/Users/jrls/Downloads/cecilia5-5.4.0/Resources/CeciliaPlot.py:977: SyntaxWarning: "is not" with a literal. Did you mean "!="?
  if self._ySpec is not 'none':
/Users/jrls/Downloads/cecilia5-5.4.0/Resources/CeciliaPlot.py:1496: SyntaxWarning: "is not" with a literal. Did you mean "!="?
  if self._xSpec is not 'none':
/Users/jrls/Downloads/cecilia5-5.4.0/Resources/CeciliaPlot.py:1519: SyntaxWarning: "is not" with a literal. Did you mean "!="?
  if self._ySpec is not 'none':

Traceback (most recent call last):
  File "/Users/jrls/Downloads/cecilia5-5.4.0/Cecilia5.py", line 115, in <module>
    sp = CeciliaSplashScreen(None, img=CeciliaLib.ensureNFD(SPLASH_FILE_PATH), callback=onStart)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jrls/Downloads/cecilia5-5.4.0/Resources/splash.py", line 53, in __init__
    self.w, self.h = self.bmp.GetWidth(), self.bmp.GetHeight()
                     ^^^^^^^^^^^^^^^^^^^
wx._core.wxAssertionError: C++ assertion ""IsOk()"" failed at /private/var/folders/w4/hp1my1ln4216vrmvp2_w45dm0000gn/T/pip-req-build-i0obld2n/ext/wxWidgets/src/osx/core/bitmap.cpp(1407) in GetWidth(): invalid bitmap

**Comment by marzaugen** [comment]
sorry, I'm not a programmer... so for me it's a bit hard to follow you. 
But I have been a total fan of Cecilia for 20 years.
is anyone able to create a patch to let Cecilia run on modern macs? I think many of us will be thankful...

**Comment by liambratchford** [comment]
Hiya, I love Cecilia and am very keen to get it back working again. I've done all the steps above , but now getting the following error, I'm new to Terminal so apologies on advance.

cd cecilia5
./Cecilia5.py
Traceback (most recent call last):
  File "/Users/liambratchford/git/cecilia5/git/cecilia5/./Cecilia5.py", line 25, in <module>
    from Resources.Variables import loadBitmaps
  File "/Users/liambratchford/git/cecilia5/git/cecilia5/Resources/Variables.py", line 24, in <module>
    from pyo import pa_get_default_devices_from_host
  File "/Users/liambratchford/Library/Python/3.9/lib/python/site-packages/pyo/__init__.py", line 27, in <module>
    from .lib import analysis as analysis
  File "/Users/liambratchford/Library/Python/3.9/lib/python/site-packages/pyo/lib/analysis.py", line 30, in <module>
    from ._core import *
  File "/Users/liambratchford/Library/Python/3.9/lib/python/site-packages/pyo/lib/_core.py", line 47, in <module>
    from .._pyo import *
ImportError: dlopen(/Users/liambratchford/Library/Python/3.9/lib/python/site-packages/pyo/_pyo.cpython-39-darwin.so, 0x0002): Library not loaded: /opt/homebrew/opt/flac/lib/libFLAC.12.dylib
  Referenced from: <E6408A3F-C3F8-30CD-906D-54CA42ADE3E5> /Users/liambratchford/Library/Python/3.9/lib/python/site-packages/pyo/libsndfile.1.dylib
  Reason: tried: '/opt/homebrew/opt/flac/lib/libFLAC.12.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/opt/flac/lib/libFLAC.12.dylib' (no such file), '/opt/homebrew/opt/flac/lib/libFLAC.12.dylib' (no such file), '/usr/local/lib/libFLAC.12.dylib' (no such file), '/usr/lib/libFLAC.12.dylib' (no such file, not in dyld cache), '/opt/homebrew/Cellar/flac/1.5.0/lib/libFLAC.12.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/Cellar/flac/1.5.0/lib/libFLAC.12.dylib' (no such file), '/opt/homebrew/Cellar/flac/1.5.0/lib/libFLAC.12.dylib' (no such file), '/usr/local/lib/libFLAC.12.dylib' (no such file), '/usr/lib/libFLAC.12.dylib' (no such file, not in dyld cache)
liambratchford@Mac cecilia5 %

**Comment by IbanEpiJos76** [comment]
> Hiya, I love Cecilia and am very keen to get it back working again. I've done all the steps above , but now getting the following error, I'm new to Terminal so apologies on advance. Any ideas?
> 
> cd cecilia5 ./Cecilia5.py Traceback (most recent call last): File "/Users/liambratchford/git/cecilia5/git/cecilia5/./Cecilia5.py", line 25, in from Resources.Variables import loadBitmaps File "/Users/liambratchford/git/cecilia5/git/cecilia5/Resources/Variables.py", line 24, in from pyo import pa_get_default_devices_from_host File "/Users/liambratchford/Library/Python/3.9/lib/python/site-packages/pyo/**init**.py", line 27, in from .lib import analysis as analysis File "/Users/liambratchford/Library/Python/3.9/lib/python/site-packages/pyo/lib/analysis.py", line 30, in from ._core import * File "/Users/liambratchford/Library/Python/3.9/lib/python/site-packages/pyo/lib/_core.py", line 47, in from .._pyo import * ImportError: dlopen(/Users/liambratchford/Library/Python/3.9/lib/python/site-packages/pyo/_pyo.cpython-39-darwin.so, 0x0002): Library not loaded: /opt/homebrew/opt/flac/lib/libFLAC.12.dylib Referenced from: /Users/liambratchford/Library/Python/3.9/lib/python/site-packages/pyo/libsndfile.1.dylib Reason: tried: '/opt/homebrew/opt/flac/lib/libFLAC.12.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/opt/flac/lib/libFLAC.12.dylib' (no such file), '/opt/homebrew/opt/flac/lib/libFLAC.12.dylib' (no such file), '/usr/local/lib/libFLAC.12.dylib' (no such file), '/usr/lib/libFLAC.12.dylib' (no such file, not in dyld cache), '/opt/homebrew/Cellar/flac/1.5.0/lib/libFLAC.12.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/Cellar/flac/1.5.0/lib/libFLAC.12.dylib' (no such file), '/opt/homebrew/Cellar/flac/1.5.0/lib/libFLAC.12.dylib' (no such file), '/usr/local/lib/libFLAC.12.dylib' (no such file), '/usr/lib/libFLAC.12.dylib' (no such file, not in dyld cache) liambratchford@Mac cecilia5 %


I've made it all the way here and found out that the specific library that isn't found by the sofware is located, from the Macintosh HD folder, in this folder --> /opt/homebrew/opt/flac/lib/ and here you find the file libFLAC.**14**.dylib. I just changed the 14 to 12 in the file name - as explained by Lawrence Whitney [here](https://www.complexityexplorer.org/courses/201-music-computation-and-complexity/forums/tutorials-technical-issues#gsc.tab=0) - and then re-executed the file with the command ./Cecilia5.py and it ran properly

**Comment by orbsmiv** [comment]
> 2\. Downgrade to Python 3.11 (ensures compatibility with the pyo module, as it provides pre-built wheels for this version)

This isn't strictly required – I'd recommend creating a `uv` virtual environment and specifying the Python version:

```
uv venv --python 3.11 cecilia-venv
source cecilia-venv/bin/activate
uv pip install pyo wxPython numpy
```

(See https://docs.astral.sh/uv/pip/environments/ for further info.)

EDIT: The above approach, i.e. allowing a `uv` provided Python, doesn't work if you want to build Cecilia5 using `py2app`. There's another approach to this, which I'll document presently.

Also, for the missing `libFLAC.12.dylib` (assuming you've installed FLAC with `brew install flac`), you can be a bit less destructive by creating a symbolic link, e.g:

```
cd /opt/homebrew/opt/flac/lib/
ln -s libFLAC.14.dylib libFLAC.12.dylib
```

**Comment by orbsmiv** [comment]
I've tested the following build process on a clean macOS Sequoia install (using [UTM](https://mac.getutm.app)):

## Prerequisites

- Homebrew (https://brew.sh/).
    - This will also download "Command Line Tools for Xcode", if required.
- `uv` (https://docs.astral.sh/uv/getting-started/installation/)

## Process

```
# We need Python 3.11 so that we don't have to compile pyo
# See https://pypi.org/project/pyo/#files for list of compiled wheels
brew install python@3.11

# Install Cecilia5 dependencies with Homebrew
brew install flac libogg libvorbis libopusenc mpg123 lame portmidi

# Cecilia5 wants libFLAC.12.dylib so we make a symlink
pushd "$(brew --prefix flac)/lib" && ln -s libFLAC.14.dylib libFLAC.12.dylib && popd

# Clone the Cecilia5 repository
git clone https://github.com/belangeo/cecilia5.git
cd cecilia5

# Create a Python virtual environment that uses the Homebrew-provided Python 3.11
uv venv --no-managed-python --python 3.11 .venv

# Activate the virtual environment
source .venv/bin/activate

# Install older setuptools - see https://github.com/ronaldoussoren/py2app/issues/531
uv pip install setuptools==70.3.0

# Install remaining Python dependencies
uv pip install pyo wxPython numpy jaraco.text py2app

# Build Cecilia5.app (in ./dist directory)
python setup.py py2app
```
